### ISSUE TITLE
MatchFirst/Or with LineStart().leaveWhitespace() has unexpected results

### ISSUE THREAD

**Issue by PhasecoreX** [pyparsing/pyparsing#272]
I am using version 2.4.7 installed from pip. It seems like `LineStart().leaveWhitespace()` doesn't want to play nicely with `MatchFirst`/`Or`.

I have simplified what I am trying to do below:
```python
from pyparsing import (
    LineEnd,
    LineStart,
    Word,
    alphas,
)

instring = """
a
 b
  c
d
e
 f
  g
"""

alpha_line = LineStart().leaveWhitespace() + Word(alphas) + LineEnd()

test = alpha_line

print(test.searchString(instring))

test2 = alpha_line | Word("_")

print(test2.searchString(instring))

test3 = alpha_line | alpha_line

print(test3.searchString(instring))
```
With `alpha_line`, I want to match lines that only have one word, with optional whitespace (spaces/tabs) before it. Testing that (first print statement) seems to work just fine. In my actual code it will be more complex than `Word(alphas)`, but it will start with `LineStart().leaveWhitespace()` and end with `LineEnd()`. Basically, I want to match an entire line ignoring leading whitespace in the line.

However when I `MatchFirst`/`Or` anything to this, as seen with `test2`, it seems to have the results thrown off (second print statement).

Even `MatchFirst`-ing `alpha_line` to itself in `test3` it doesn't work (third print statement).

The output of these are:
```
[['a', '\n'], ['b', '\n'], ['c', '\n'], ['d', '\n'], ['e', '\n'], ['f', '\n'], ['g', '\n']]
[['a', '\n'], ['d', '\n'], ['e', '\n']]
[['a', '\n'], ['d', '\n'], ['e', '\n']]
```
I assume I am doing something wrong, or not understanding completely how `LineStart().leaveWhitespace()` works, but I also wouldn't assume that adding `| Word("_")` would alter the results at all. I figured in `test2` the MatchFirst of `alpha_line` would happen first every time. It's also interesting to me that the `d` line "fixes" it so it starts matching again (until spaces show up on line `f`).

**Comment by ptmcg** [comment]
I have been poking at this problem for the past month or so, and have finally come up with a resolution.

LineStart and StringStart have gotten tightened up definitions, in that they only succeed if currently at the start of a line or start of a string without skipping whitespace. So for your example string, you will only get matches for lines starting with 'a', 'd', and 'e'. With the latest fixes, you will get the same results regardless of whether the expression starting with LineStart() is alone, part of an And, MatchFirst, Group, or other expression.

Thanks so much for submitting this, I think this really improves a feature in pyparsing that has been "iffy" in the past.

This will ship out in the 3.0.0rc2 release, later this week.

**Comment by ptmcg** [comment]
Released in 3.0.0rc2
